locator/abstract_replication_strategy.cc

```diff
@@ -169,6 +169,15 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(cached_endpoint_set& cached_eps, const token& search_token, const token_metadata& tm) const {
+    auto cur_ring_version = tm.get_ring_version();
+    assert(cur_ring_version >= 0);
```

Comment:
I'm confused. token_metadata_ptr already is the embodiment of a ring topology mixed into a replication strategy. Here you are replicating that mechanism with another cache on top of it.

I don't understand what's the goal of maybe_calculate_natural_endpoints. It looks like it's optimizing the easy case and rare cases, local or everywhere, and leaving the common and hard cases unoptimized.

Reply:
The goal is indeed to optimize the local and everywhere cases while keeping the other cases unoptimized.
`abstract_replication_strategy::calculate_natural_endpoints` is meant to be called on arbitrary token_metadata.
We use it typically on token_metadata that was cloned and possibly manipulated so it doesn't belong to any effective_replication_map - hence the natural endpoints for it aren't already precalculated.

In this patch `maybe_calculate_natural_endpoints` is used only internally by `abstract_replication_strategy`.
But I plan to use it in repair based node ops.  Maybe they should build an effective_replication_map (that is aware of this optimization) and use that instead of calling the abstract_replication_strategy calculate_natural_endpoints.

Reply:
@avikivity I don't want this last patch hold up the whole series.
Are you ok with defining the method private to abstract_replication_strategy and use it only internally there for now?

Reply:
Sorry, I lost the thread. How about you drop this last patch and resurrect it in a separate series? I think it's not crucial for what the series does.

Reply:
done in v7